File: src/test/java/no/nav/pensjon/selvbetjeningopptjening/util/PeriodeUtilTest.java
package no.nav.pensjon.selvbetjeningopptjening.util;

import no.nav.pensjon.selvbetjeningopptjening.opptjening.Periode;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class PeriodeUtilTest {

    @Test
    void test_that_sortPerioderByFomDate_sorts_by_fom_date() {
        List<TestPeriode> actual = PeriodeUtil.sortPerioderByFomDate(
                List.of(
                        new TestPeriode(
                                LocalDate.of(2020, 1, 1),
                                LocalDate.of(2020, 1, 31)),
                        new TestPeriode(
                                LocalDate.of(2019, 1, 1),
                                LocalDate.of(2019, 1, 31))));

        assertEquals(2019, actual.get(0).getFomDato().getYear());
        assertEquals(2020, actual.get(1).getFomDato().getYear());
    }

    @Test
    void test_that_sortPerioderByFomDate_disregards_tom_date() {
        List<TestPeriode> actual = PeriodeUtil.sortPerioderByFomDate(
                List.of(
                        new TestPeriode(
                                LocalDate.of(2020, 1, 1),
                                LocalDate.of(2020, 1, 31)),
                        new TestPeriode(
                                LocalDate.of(2019, 1, 1),
                                LocalDate.of(2020, 2, 1))));

        assertEquals(2019, actual.get(0).getFomDato().getYear());
        assertEquals(2020, actual.get(1).getFomDato().getYear());
    }

    @Test
    void when_periods_equal_then_isPeriodeWithinInterval_returns_false() {
        boolean actual = PeriodeUtil.isPeriodeWithinInterval(
                new TestPeriode(
                        LocalDate.of(2020, 1, 1),
                        LocalDate.of(2020, 1, 31)),
                LocalDate.of(2020, 1, 1),
                LocalDate.of(2020, 1, 31));

        assertFalse(actual);
    }

    @Test
    void when_period_outside_interval_on_both_sides_then_isPeriodeWithinInterval_returns_false() {
        boolean actual = PeriodeUtil.isPeriodeWithinInterval(
                new TestPeriode(
                        LocalDate.of(2020, 1, 1),
                        LocalDate.of(2020, 1, 31)),
                LocalDate.of(2020, 1, 2),
                LocalDate.of(2020, 1, 30));

        assertFalse(actual);
    }

    @Test
    void when_period_outside_interval_on_start_side_then_isPeriodeWithinInterval_returns_false() {
        boolean actual = PeriodeUtil.isPeriodeWithinInterval(
                new TestPeriode(
                        LocalDate.of(2020, 1, 1),
                        LocalDate.of(2020, 1, 30)),
                LocalDate.of(2020, 1, 2),
                LocalDate.of(2020, 1, 31));

        assertFalse(actual);
    }

    @Test
    void when_period_outside_interval_on_end_side_then_isPeriodeWithinInterval_returns_false() {
        boolean actual = PeriodeUtil.isPeriodeWithinInterval(
                new TestPeriode(
                        LocalDate.of(2020, 1, 2),
                        LocalDate.of(2020, 1, 31)),
                LocalDate.of(2020, 1, 1),
                LocalDate.of(2020, 1, 30));

        assertFalse(actual);
    }

    @Test
    void when_period_strictly_within_inteval_then_isPeriodeWithinInterval_returns_true() {
        boolean actual = PeriodeUtil.isPeriodeWithinInterval(
                new TestPeriode(
                        LocalDate.of(2020, 1, 2),
                        LocalDate.of(2020, 1, 30)),
                LocalDate.of(2020, 1, 1),
                LocalDate.of(2020, 1, 31));

        assertTrue(actual);
    }

    private static class TestPeriode implements Periode {

        private final LocalDate start;
        private final LocalDate end;

        private TestPeriode(LocalDate start, LocalDate end) {
            this.start = start;
            this.end = end;
        }

        @Override
        public LocalDate getFomDato() {
            return start;
        }

        @Override
        public LocalDate getTomDato() {
            return end;
        }
    }
}

File: src/test/java/no/nav/pensjon/selvbetjeningopptjening/opptjening/PeriodeUtilTest.kt
package no.nav.pensjon.selvbetjeningopptjening.opptjening

import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.shouldBe
import java.time.LocalDate

class PeriodeUtilTest : ShouldSpec({

    context("sortPerioderByFomDate") {
        should("sort by f.o.m.-dato") {
            val actual: List<TestPeriode> = PeriodeUtil.sortPerioderByFomDate(
                perioder = listOf(
                    TestPeriode(
                        start = LocalDate.of(2020, 1, 1),
                        end = LocalDate.of(2020, 1, 31)
                    ),
                    TestPeriode(
                        start = LocalDate.of(2019, 1, 1),
                        end = LocalDate.of(2019, 1, 31)
                    )
                )
            )

            actual[0].getFomDato().year shouldBe 2019
            actual[1].getFomDato().year shouldBe 2020
        }

        should("disregard t.o.m.-dato") {
            val actual: List<TestPeriode> = PeriodeUtil.sortPerioderByFomDate(
                perioder = listOf(
                    TestPeriode(
                        start = LocalDate.of(2020, 1, 1),
                        end = LocalDate.of(2020, 1, 31)
                    ),
                    TestPeriode(
                        start = LocalDate.of(2019, 1, 1),
                        end = LocalDate.of(2020, 2, 1)
                    )
                )
            )

            actual[0].getFomDato().year shouldBe 2019
            actual[1].getFomDato().year shouldBe 2020
        }
    }

    context("isPeriodeWithinInterval") {
        should("give 'false' when the period is equal to the interval") {
            PeriodeUtil.isPeriodeWithinInterval(
                periode = TestPeriode(
                    start = LocalDate.of(2020, 1, 1),
                    end = LocalDate.of(2020, 1, 31)
                ),
                start = LocalDate.of(2020, 1, 1),
                end = LocalDate.of(2020, 1, 31)
            ) shouldBe false
        }

        should("give 'false' when the period is outside the interval on both sides") {
            PeriodeUtil.isPeriodeWithinInterval(
                periode = TestPeriode(
                    start = LocalDate.of(2020, 1, 1),
                    end = LocalDate.of(2020, 1, 31)
                ),
                start = LocalDate.of(2020, 1, 2),
                end = LocalDate.of(2020, 1, 30)
            ) shouldBe false
        }

        should("give 'false' when the period is outside the interval on start side") {
            PeriodeUtil.isPeriodeWithinInterval(
                periode = TestPeriode(
                    start = LocalDate.of(2020, 1, 1),
                    end = LocalDate.of(2020, 1, 30)
                ),
                start = LocalDate.of(2020, 1, 2),
                end = LocalDate.of(2020, 1, 31)
            ) shouldBe false
        }

        should("give 'false' when the period is outside the interval on end side") {
            PeriodeUtil.isPeriodeWithinInterval(
                periode = TestPeriode(
                    start = LocalDate.of(2020, 1, 2),
                    end = LocalDate.of(2020, 1, 31)
                ),
                start = LocalDate.of(2020, 1, 1),
                end = LocalDate.of(2020, 1, 30)
            ) shouldBe false
        }

        should("give 'true' when the period is strictly within the interval") {
            PeriodeUtil.isPeriodeWithinInterval(
                periode = TestPeriode(
                    start = LocalDate.of(2020, 1, 2),
                    end = LocalDate.of(2020, 1, 30)
                ),
                start = LocalDate.of(2020, 1, 1),
                end = LocalDate.of(2020, 1, 31)
            ) shouldBe true
        }
    }

})

data class TestPeriode(
    val start: LocalDate,
    val end: LocalDate
) : Periode {
    override fun getFomDato(): LocalDate = start

    override fun getTomDato(): LocalDate = end
}
